### PR TITLE
api integration with axios

### PR DESCRIPTION
**Pull request by milan0827**
A simple UI and performing CRUD operation using axios.
json-server is a package manager for creating simple REST API which was used for this project to perofrom CRUD operation and testing purpose.

This is a simple user form 👇
Screenshot 2024-01-29 203154

### CODE REVIEW

File: src/services/apiServices.ts
Comment (sarojghalan) [inline src/services/apiServices.ts:1]: its a nice process, but how would you handle `error` and `success` message. 

creating an instance of `axios` would be nice, https://axios-http.com/docs/instance. 

and whenever you want which request can be handled easily with,

File: src/pages/UserForm.tsx/index.tsx [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+      firstName: data.firstName,
+      lastName: data.lastName,
+      email: data.email,
+      age: data.age,
+      address: data.address,
+      country: data.country,
+      city: data.city,
+      gender: data.gender,
+      martialStatus: data.martialStatus,
+    };
```
Comment (sarojghalan) [inline src/pages/UserForm.tsx/index.tsx]:
```suggestion
 createUser(...data)
```

File: src/pages/UserForm.tsx/index.tsx [hunk 2]
```diff
@@ -0,0 +1,165 @@
+import { useState } from "react";
+import { createUser } from "../../services/apiServices";
+import { UserDataType } from "../../shared/type";
+
+function UserForm() {
+  const [userData, setUserData] = useState<UserDataType[]>([]);
```
Comment (sarojghalan) [inline src/pages/UserForm.tsx/index.tsx]: is this is used or ?

**Review by sarojghalan — CHANGES_REQUESTED** [review]
nice your logic are really increasing, 
but still minor changes are need to me fix

File: src/services/apiServices.ts
Comment (milan0827) [reply src/services/apiServices.ts:1]: I will researching more on axios instance,

File: src/pages/UserForm.tsx/index.tsx (hunk 1 above)
Comment (milan0827) [reply src/pages/UserForm.tsx/index.tsx]: const newData = {
      firstName: data.firstName,
      lastName: data.lastName,
      email: data.email,
      age: data.age,
      address: data.address,
      country: data.country,
      city: data.city,
      gender: data.gender,
      martialStatus: data.martialStatus,
    };
As these were unnecessary, so these lines of codes were removed

File: src/pages/UserForm.tsx/index.tsx (hunk 2 above)
Comment (milan0827) [reply src/pages/UserForm.tsx/index.tsx]: No it was not used and removed in the updated code

File: src/pages/UserForm.tsx/index.tsx [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+    martialStatus: "",
+  });
+
+  function handleInputChange(e: React.ChangeEvent<HTMLInputElement>) {
+    setData({ ...data, [e.target.name]: e.target.value });
+  }
+
+  function handleCreateData(e: React.ChangeEvent<HTMLFormElement>) {
+    e.preventDefault();
+    createUser(data);
```
Comment (sarojghalan) [inline src/pages/UserForm.tsx/index.tsx]: handle success and error message

File: src/pages/UserForm.tsx/index.tsx
Comment (sarojghalan) [inline src/pages/UserForm.tsx/index.tsx:1]: i think validation is missing, handle validation precisely

File: src/pages/UserForm.tsx/index.tsx [hunk 4]
```diff
@@ -0,0 +1,141 @@
+import { useState } from "react";
+import { createUser } from "../../services/apiServices";
+import { UserDataType } from "../../shared/type";
```
Comment (sarojghalan) [inline src/pages/UserForm.tsx/index.tsx]: handle import with absolute import.

File: src/pages/UserForm.tsx/index.tsx [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+          type="text"
+          placeholder="Martial status"
+          className="input"
+          name="martialStatus"
+          value={data.martialStatus}
+          onChange={handleInputChange}
+        />
+      </div>
+
+      <button className="bg-blue-400 p-4 text-gray-200">Submit</button>
```
Comment (sarojghalan) [inline src/pages/UserForm.tsx/index.tsx]:
```suggestion
      <button type="submit" className="bg-blue-400 p-4 text-gray-200">Submit</button>
```

File: src/services/apiServices.ts
Comment (sarojghalan) [inline src/services/apiServices.ts:1]: implement , `interceptor` to handle our request before `request` and `response` .

**Review by sarojghalan — CHANGES_REQUESTED** [review]
nice, a lot of new features has been implemented ,

still minor changes are required in new features :rose:

File: src/services/apiServices.ts [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+
+export async function geAlltUser() {
+  const data = await userApi.get("/");
+  return data;
+}
+
+export async function getUser(id: string) {
+  const data = await userApi.get(`/${id}`);
+  return data;
+}
```
Comment (sarojghalan) [inline src/services/apiServices.ts]: i think these are now hooks now you can separate it. 

as our `apiServices` will only have context about network .

Comment (sarojghalan) [reply src/services/apiServices.ts]: research on where you should place hooks

**Review by sarojghalan — CHANGES_REQUESTED** [review]
nice bhai, Surra :smile: 
but it got comments :smiley:

File: src/index.css [hunk 7]
```diff
@@ -4,6 +4,11 @@
 
 @layer components {
   .input {
```
Comment (sarojghalan) [inline src/index.css:6]: what if we have another input but with different ui structure.

File: src/pages/TableView/index.tsx [hunk 8]
```diff
@@ -1,6 +1,44 @@
...
+  const [userData, setUserData] = useState([]);
+
+  useEffect(function () {
+    getAllUser().then((data: UserDataType[]) => setUserData(data));
+  }, []);
+
   // Data to be displayed
-  return <div>TableView</div>;
+  return (
+    <>
```
Comment (sarojghalan) [inline src/pages/TableView/index.tsx:27]: wrapping it with section main tag would be nice.

File: src/pages/UserForm.tsx/index.tsx
Comment (sarojghalan) [reply src/pages/UserForm.tsx/index.tsx:1]: validation missing

File: src/services/apiServices.ts
Comment (sarojghalan) [reply src/services/apiServices.ts:1]: we will talk about this tomorrow.

File: src/utils/helpers.ts
Comment (sarojghalan) [inline src/utils/helpers.ts:1]: we will talk about this file, tomorrow

**Review by sarojghalan — CHANGES_REQUESTED** [review]
nice bhai, 

still minor changes are need to be done.

File: src/hooks/useDelete.tsx [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+
+  async function deleteUser() {
+    setIsLoading(true);
+    try {
+      console.log(url, data.id, "url-path");
+      await service.delete(`${url}${data.id}`).then((data) => {
+        setUserData(userList.filter((datas) => data.data.id !== datas.id));
+      });
+      setIsLoading(false);
+      return data;
```
Comment (sarojghalan) [inline src/hooks/useDelete.tsx]: is returning data necessary?

File: src/pages/TableView/index.tsx [hunk 10]
```diff
@@ -0,0 +1,71 @@
+import useFetch from "../../hooks/useFetch";
+import { UserDataType } from "../../shared/type";
+import TableRow from "./TableRow";
+import { useEffect, useState } from "react";
+import { useNavigate } from "react-router-dom";
+
+const BASE_URL = "http://localhost:3000/user";
```
Comment (sarojghalan) [inline src/pages/TableView/index.tsx]: we have, base url on service so why this is here

File: src/pages/UserForm.tsx/index.tsx [hunk 11]
```diff
@@ -0,0 +1,267 @@
...
+  // userDataValues.forEach((el, i) => {
+
+  // });
+
+  const HandleCreateData = (e: React.ChangeEvent<HTMLFormElement>) => {
+    e.preventDefault();
+
+    const userDataValues = Object.values(userData);
+
+    for (let i = 0; i < userDataValues.length; i++) {
```
Comment (sarojghalan) [inline src/pages/UserForm.tsx/index.tsx:66]: we will talk on this tomorrow

**Review by sarojghalan — CHANGES_REQUESTED** [review]
Nice  now, Research on `zod` ,
its a validation helper
Still minor changes are need to fix

File: src/pages/TableView/index.tsx (hunk 8 above)
Comment (milan0827) [reply src/pages/TableView/index.tsx:27]: yes, you are correct

File: src/pages/UserForm.tsx/index.tsx (hunk 3 above)
Comment (milan0827) [reply src/pages/UserForm.tsx/index.tsx]: All success and error are set dai

File: src/pages/UserForm.tsx/index.tsx
Comment (milan0827) [reply src/pages/UserForm.tsx/index.tsx:1]: I have done all the validation, you can review it.

File: src/services/apiServices.ts (hunk 6 above)
Comment (milan0827) [reply src/services/apiServices.ts]: I have created the custom hooks and placed it separate files.

File: src/index.css (hunk 7 above)
Comment (milan0827) [reply src/index.css:6]: I hvae to research on this one.

File: src/hooks/useDelete.tsx (hunk 9 above)
Comment (milan0827) [reply src/hooks/useDelete.tsx]: Actually it's not and I hvae removed.

File: src/pages/TableView/index.tsx (hunk 10 above)
Comment (milan0827) [reply src/pages/TableView/index.tsx]: you are right daii and I have removed.

**Comment by milan0827** [comment]
Thanks @sarojghalan  daiii for your amazing guidance

File: src/pages/UserForm.tsx/index.tsx (hunk 4 above)
Comment (milan0827) [reply src/pages/UserForm.tsx/index.tsx]: Yes, I have done it, you can review my code.

File: src/hooks/useFetch.tsx
Comment (sarojghalan) [inline src/hooks/useFetch.tsx:1]: nice, bhai

File: src/pages/UserForm.tsx/index.tsx
Comment (sarojghalan) [inline src/pages/UserForm.tsx/index.tsx:1]: is validation still missing or what ?

File: src/routes/routes.tsx
Comment (sarojghalan) [inline src/routes/routes.tsx:1]: you have define routes, as of now we do not need loader, try to implement loader too for future easiness, 

lazy, loading - follow official react router docs

**Review by sarojghalan — CHANGES_REQUESTED** [review]
nice bhai, you are rocking 

keep up the pace :rose:

File: src/hooks/useFetch.tsx
Comment (milan0827) [reply src/hooks/useFetch.tsx:1]: Dhanyabaad daii

File: src/routes/routes.tsx
Comment (milan0827) [reply src/routes/routes.tsx:1]: okay, I will research on that.

File: src/pages/UserForm.tsx/index.tsx
Comment (milan0827) [reply src/pages/UserForm.tsx/index.tsx:1]: No, It's not missing, All the validation are working fine.

**Pull request merged**